module/zfs/spa_errlog.c

```diff
@@ -128,6 +241,300 @@ spa_log_error(spa_t *spa, const zbookmark_phys_t *zb)
 	mutex_exit(&spa->spa_errlist_lock);
 }
 
+#ifdef _KERNEL
+static int
+find_block_txg(dsl_dataset_t *ds, zbookmark_err_phys_t *zep,
+    uint64_t *birth_txg)
+{
+	objset_t *os;
+	int error = dmu_objset_from_ds(ds, &os);
+	if (error != 0)
+		return (error);
+
+	dnode_t *dn;
+	blkptr_t bp;
+
+	error = dnode_hold(os, zep->zb_object, FTAG, &dn);
+	if (error != 0)
+		return (error);
+
+	rw_enter(&dn->dn_struct_rwlock, RW_READER);
+	error = dbuf_dnode_findbp(dn, zep->zb_level, zep->zb_blkid, &bp, NULL,
+	    NULL);
+
+	if (error == 0 && BP_IS_HOLE(&bp))
+		error = SET_ERROR(ENOENT);
+
+	*birth_txg = bp.blk_birth;
+	rw_exit(&dn->dn_struct_rwlock);
+	dnode_rele(dn, FTAG);
+	return (error);
+}
+
+/*
+ * This function serves a double role. If only_count is true, it returns
+ * how many times an error block belonging to this filesystem is referenced by
+ * snapshots or clones. If only_count is false, each time the error block is
+ * referenced by a snapshot or clone, it returns a pointer to the bookmark of
+ * the error block back to process_error_log().
+ */
+static int
+check_filesystem(spa_t *spa, uint64_t fs, zbookmark_err_phys_t *zep,
+    uint64_t *count, void *addr, boolean_t only_count)
+{
+	dsl_dataset_t *ds;
+	dsl_pool_t *dp = spa->spa_dsl_pool;
+
+	int error = dsl_dataset_hold_obj(dp, fs, FTAG, &ds);
+	if (error != 0)
+		return (error);
+
+	uint64_t latest_txg;
+	uint64_t txg_to_consider = spa->spa_syncing_txg;
+	boolean_t check_snapshot = B_TRUE;
+	error = find_block_txg(ds, zep, &latest_txg);
+	if (error == 0) {
+		if (zep->zb_birth < latest_txg) {
+			txg_to_consider = latest_txg;
+		} else {
```

Comment:
in this case should zb_birth==latest_txg, i.e. the block in this snapshot is the same as the block that had the error?  Maybe the check should be 
```c
if (latest_txg == zep->zb_birth) {
    ... add to count
} else {
...
}